### ISSUE TITLE
`metatensor-learn`: next steps

### ISSUE THREAD

**Issue by jwa7** [lab-cosmo/metatensor#446]
Post (almost merged) Dataset and DataLoader, let's define next steps for `metatensor-learn`:

- [ ] A DataLoader that wraps DataLoaders, for learning on multiple targets and/or Datasets. See: https://github.com/lab-cosmo/metatensor-models/pull/27 in `metatensor-models` - there is probably a nice way of absorbing this generically into `learn`.
- [ ] Implement `ModuleMap`
- [ ] Loss functions - start with a simple module that calculates the L2 loss between 2 TensorMaps (I have one in `rho_learn` that we could use as a starting point: https://github.com/jwa7/rho_learn/blob/main/rholearn/loss.py). Plenty of API considerations here: reduction methods, returning global loss, loss per sample, handling sparsity, etc etc

**Comment by agoscinski** [comment]
## metatensor-operations required for solvers in linear and kernel models

- [ ] solve_cholesky (this should be the same doing Cholesky decomposition + triangular solve as done in the kernel)
      https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cho_solve.html
      https://pytorch.org/docs/stable/generated/torch.cholesky_solve.html
- [ ] eigh
      https://numpy.org/doc/stable/reference/generated/numpy.linalg.eigh.html
      https://pytorch.org/docs/stable/generated/torch.linalg.eigh.html

**Comment by agoscinski** [comment]
Follow up Issues from PR https://github.com/lab-cosmo/metatensor/pull/427:
- [x] merge the construction of test tensor maps https://github.com/lab-cosmo/metatensor/issues/451 to then add tests with multiple blocks
- [x] Figure out to add forward function into doc that is TorchScript compatible: TorchScript allows not # added to appear in doc, :inherited-members: is not compatible with torch
But also does not allow to use super in forward https://github.com/pytorch/pytorch/issues/42885
- [ ] add weight and bias  property to Linear (will be probably List[nn.Parameter])
- [ ] Consider to also accept single Labels for `out_properties` for `ModuleMap.from_module` and `Linear` default constructor. It is also a bit tricky to the get the list check done in a way that does not break in metatensor-torch. I was not sure if this is useful in the first place, so I skipped this.
- [ ] add examples for usage of ModuleMap as mentioned in https://github.com/lab-cosmo/metatensor/pull/427#issuecomment-1849001299

_Originally posted by @agoscinski in https://github.com/lab-cosmo/metatensor/issues/427#issuecomment-1911211939_

**Comment by jwa7** [comment]
> Figure out to add forward function into doc that is TorchScript compatible: TorchScript allows not # added to appear in doc, :inherited-members: is not compatible with torch
But also does not allow to use super in forward https://github.com/pytorch/pytorch/issues/42885

Note that in the recent PR #513 this was solved by moving to compositional modules as opposed to inherited ones
